storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3110,1746 @@ buf_block_t *fsp_get_latched_xdes_page(
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_foreign_cols);
+    if (err == DB_SUCCESS)
+      err= defragment_table(dict_sys.sys_virtual);
+    return err;
+  }
+
+  /** @return extent which replaces the later extent
+  or same extent if there is no replacement exist */
+  uint32_t get_new_extent(uint32_t old_extent) const noexcept
+  {
+    auto it= m_extent_map.find(old_extent);
+    if (it != m_extent_map.end())
+      return it->second;
+    return old_extent;
+  }
+
+  /** @return state for the given extent */
+  uint32_t get_state(uint32_t extent) noexcept
+  {
+    return m_extent_info[extent];
+  }
+};
+
+/** Validate the system tablespace list */
+__attribute__((warn_unused_result))
+static dberr_t fsp_tablespace_validate(fil_space_t *space,
+                                       mtr_t *mtr) noexcept
+{
+  /* Validate all FSP list in system tablespace */
+  dberr_t err= DB_SUCCESS;
+  if (buf_block_t *header= fsp_get_header(space, mtr, &err))
+  {
+    err= flst_validate(header, FSP_FREE + FSP_HEADER_OFFSET, mtr);
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_FREE_FRAG + FSP_HEADER_OFFSET, mtr);
+
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_FULL_FRAG, mtr);
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL, mtr);
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE, mtr);
+  }
+  return err;
+}
+
+dberr_t IndexDefragmenter::store_child_pages(buf_block_t *block) noexcept
+{
+  const byte *page= block->page.frame;
+  const rec_t *rec= page_rec_get_next_low(page + PAGE_OLD_INFIMUM, false);
+  while (rec != page + PAGE_OLD_SUPREMUM)
+  {
+    ulint len;
+    ulint offset= rec_get_nth_field_offs_old(rec,
+                                             rec_get_n_fields_old(rec) - 1,
+                                             &len);
+    if (len != 4)
+      return DB_CORRUPTION;
+
+    if (offset >= srv_page_size)
+      return DB_CORRUPTION;
+
+    const byte *field= rec + offset;
+    /* m_parent_pages[child_page_no] =
+       1st 32 bit to indicate offset in parent page
+       2nd 32 bit to indicate parent page number */
+    m_parent_pages[mach_read_from_4(field)]=
+      uint64_t(page_offset(field)) << 32 | block->page.id().page_no();
+    rec= page_rec_get_next_low(rec, false);
+  }
+  return DB_SUCCESS;
+}
+
+dberr_t IndexDefragmenter::get_level_block(uint16_t level, mtr_t *mtr,
+                                           uint32_t *cur_page_no) noexcept
+{
+  uint32_t child_page_no= m_index.page;
+  dberr_t err= DB_SUCCESS;
+  while (1)
+  {
+    buf_block_t *block= fsp_get_latched_page(page_id_t{0, child_page_no},
+                                             mtr, &err);
+    if (!block)
+      return err;
+
+    page_t *page= buf_block_get_frame(block);
+    uint16_t cur_level= btr_page_get_level(page);
+    if (cur_level == level)
+      break;
```

Comment:
I don’t see any reference to the previously read level here. We must validate that the `btr_page_get_level(page)` is decreasing by one at each step. By the way, it needs to be `const page_t  *page`; we are not supposed to modify anything here.

Please refer to the variable `height` in `btr_cur_t::search_leaf()`.